### ISSUE TITLE
C Syntax Tree needs to be Updated

### ISSUE THREAD

**Issue by Raghava-Ch** [tree-sitter/tree-sitter-c#129]
C Language syntax tree hierarchy should be modified for the functions which returns pointer.
C function definition with pointer type nodes are expected to be part of the "type: primitive_type". But in current implementation pointer declarator out of the primitive type node as shown in the example below.

Because of this node entry which is out of the primitive type **common query to list all the function definitions is not able to identify it as a function definition**.

### example function definition:
`const char ** fn_returns_pointer() { }`

### parsed AST:
```
translation_unit [0, 0] - [3, 0]
  function_definition [0, 0] - [2, 1]
    type_qualifier [0, 0] - [0, 5]
    type: primitive_type [0, 6] - [0, 10]
    declarator: pointer_declarator [0, 11] - [0, 34] ---------------------
      declarator: pointer_declarator [0, 12] - [0, 34]                   | these pointer_declarators are not expected to be here
        declarator: function_declarator [0, 14] - [0, 34]-----------------
          declarator: identifier [0, 14] - [0, 32]
          parameters: parameter_list [0, 32] - [0, 34]
    body: compound_statement [1, 0] - [2, 1]
```

**Comment by XVilka** [comment]
I think we met this situation too, while debugging these two issues:
- https://github.com/rizinorg/rizin/issues/2190
- https://github.com/rizinorg/rizin/pull/2194
- https://github.com/rizinorg/rizin/issues/1272

**Comment by touzeauv** [comment]
I just had a look at the grammar, intending to fix this issue (and issue #126 which I think is related).

After exploring a bit the grammar and the C syntax, I believe the behavior of tree-sitter is actually correct in the sense that it produces a concrete syntax tree that is closer to what the syntax expressed.

I think it is useful to consider the following example:
`int* x, y;`
In C, this declares a pointer (x) to an int, and an int (y), i.e. the "star" only apply to x.

Although this is very uncommon, it is possible to use a similar syntax to declare functions:
`int* f(), g();
// Equivalent to
// int* f();
// int g();  <- Note the return type
`

I thus believe grouping the "star" with the type specifier is incorrect because the type specifier might be common to several declarators. Having one type specifier per declarator might be useful and more intuitive, but the result would be closer to an AST than a CST.

**Comment by Raghava-Ch** [comment]
In fact I realised what @touzeauv is correct, since tree-sitter is intended to use it as syntax highlighter.

But I am using tree-sitter for the code analysis since it has flexible license and easy to expand for the other languages. Since my requirements are a little different than the actual tree-sitter purpose.

Anyway there is no other library like tree-sitter, If it can serve the complete computer languages requirements it would be great.
